### PR TITLE
[Stdlib][Build] Fix the standard library build for fully static Linux.

### PR DESCRIPTION
**Pull request by al45tair**
Mostly this just means adding `Musl` as a module dependency of various things and making sure that we build things for `swift_static` even if `SWIFT_BUILD_STATIC_STDLIB` isn't enabled.

There's also a slight difference in the declaration of `memcmp()`; musl's declaration is more like the one we have on Darwin.

rdar://123508245

### CODE REVIEW

**Comment by al45tair** [comment]
This sits on top of #71847.

**Comment by al45tair** [comment]
It's just [this one changeset](https://github.com/apple/swift/pull/71848/commits/8f4a746b0784b9839cb81a9d6254d1a251f7dfa2) that needs reviewing here. The other changes come from upstream PRs that are being reviewed by others and that have already been updated, so please don't waste time looking at those.

File: stdlib/public/SwiftShims/swift/shims/CMakeLists.txt [hunk 1]
```diff
@@ -34,11 +34,9 @@ add_custom_command(
     OUTPUT "${output_dir}"
     COMMAND ${CMAKE_COMMAND} "-E" "make_directory" "${output_dir}")
 
-if(SWIFT_BUILD_STATIC_STDLIB)
-  add_custom_command(
-      OUTPUT "${output_dir_static}"
-      COMMAND ${CMAKE_COMMAND} "-E" "make_directory" "${output_dir_static}")
-endif()
+add_custom_command(
```
Comment (etcwilde) [inline stdlib/public/SwiftShims/swift/shims/CMakeLists.txt:37]: Is this meant to be completely uncovered, or should this and the non-static case be guarded on whether or not a static build is requested, or whether it's very specifically the Linux-static SDK build (sort of like the stdlib CLibdispatch build up above)?

File: utils/swift_build_support/swift_build_support/targets.py [hunk 2]
```diff
@@ -8,8 +8,11 @@
 # See https://swift.org/LICENSE.txt for license information
 # See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
+import atexit
```
Comment (etcwilde) [inline utils/swift_build_support/swift_build_support/targets.py:11]: I might be missing something, but where are these imports being used?

File: utils/swift_build_support/swift_build_support/products/llvm.py [hunk 3]
```diff
@@ -11,6 +11,7 @@
 # ----------------------------------------------------------------------------
 
 import os
+import errno
```
Comment (etcwilde) [inline utils/swift_build_support/swift_build_support/products/llvm.py:14]: Not seeing where this import is getting used?

File: utils/swift_build_support/swift_build_support/targets.py (hunk 2 above)
Comment (al45tair) [reply utils/swift_build_support/swift_build_support/targets.py:11]: That isn't part of 8f4a746, and I'm not actually certain what added it — but in any case the Python code gets linted so unused imports have to be removed for the tests to pass (I know I've removed a few of these upstream).

File: utils/swift_build_support/swift_build_support/products/llvm.py (hunk 3 above)
Comment (al45tair) [reply utils/swift_build_support/swift_build_support/products/llvm.py:14]: Again, not part of 8f4a746, and I'm pretty sure I've already removed it in an upstream PR. (I haven't rebased this one yet…).

File: stdlib/public/SwiftShims/swift/shims/CMakeLists.txt (hunk 1 above)
Comment (al45tair) [reply stdlib/public/SwiftShims/swift/shims/CMakeLists.txt:37]: Having it uncovered is pretty harmless; it does mean we generate an extra `swift_static` directory that wasn't there before, containing the shims and a symlink to the clang headers. I _guess_ we could loop over all of the configured SDKs to see if any of them are static-only, which would avoid doing that if we haven't been asked to do any kind of static stdlib build.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Pull request merged**